Make AbstractNamedBean.toString() final

danielb987 opened the pull request:
The purpose of this PR is to show the consequences of making `AbstractNamedBean.toString()` final.

Most of the toString() methods seem to be removable, but these may cause problems:
jmri.implementation.DefaultRailCom
jmri.implementation.DefaultSignalSystem
jmri.jmrit.logix.Portal

danielb987 commented:
@pabender 
I'm fine with that, but in that case the javadoc should be updated to reflect that. Currently `AbstractNamedBean.toString()` states _It would be good to eventually make this final to keep it consistent system-wide_ and some of the implementing classes states that they violates the contract in `NamedBean.toString()`.

On the other hand, the javadoc of `NamedBean.toString()` states that toString() `Display the system-specific name`. So the contract of NamedBean says that toString() should return the system name. If this is wrong, the javadoc for `NamedBean.toString()` should be updated as well.

So we have a conflict between the contract of `Object.toString()` and `NamedBean.toString()`. If Object.toString() should rule, I can create a new PR that changes the javadoc for NamedBean.toString and the affected classes that are mention in this PR.

bobjacobsen commented:
> In the specific case of IDTags and RailCom objects, the toString is the report presented by the associated reporters on both panels and in the table. By design, reporters report objects and use the toString method when asked for a the report.
> 

This part gets tricky. 

`toString` is specified as representing an _object_ not the state of that object. A lot of infrastructure code, e.g. in Swing, assumes the result of `toString` is invariant.

pabender commented:
Hmmm... 

On the one hand, we can be more restrictive than Object.toString’s contract.

On the other hand, At one point in time, a system prefix and type letter could precisely identify an object.  Since system prefixes are now arbitrary, using the system name as the output of toString provides little information in the absence of other context.

That said, my biggest issue with this change is it breaks the behavior of all of our reporters that report IDTag or RailCom objects.

If we take the system name of an RFID tag ( which is represented using an IdTag object ) as an example, the system name for the tags most of us use is based on the 20 ASCII character random value encoded into the tag itself.  Displaying the system name in this case is absolutely meaningless because we have no idea what that tag is representing.

Unless we change the contract of the Reporter so that it reports something more restrictive than Object, the toString Methods for IDTags and RailCom objects will need to remain the same.  Changing the contract for Reporters WILL require changes to LocoNet reporters, which currently report String objects.

I don’t have a strong opinion on how this impacts other named bean types.

bobjacobsen commented:
@pabender  Although I agree that user names "unlikely to change", @KenC57 has been making a pretty good case that there are cases where that change is valuable.

There are two kinds of problems with allowing beans to make run-time changes in the result of `toString()`

- (bad) the change is so large that it results in changes to the sort order of items in tables, trees, JComboBoxes and more generally the sorted Collections.   In the Swing case, this is generally disastrous: you can select the wrong item, get `ConcurrentModificationExceptions` and just generally have Bad Things happen.  We really need to avoid this.

- (not so bad) But if the sort order is not affected, as far as I know the only issue is the possibility of stale content:  An item was filled out from an earlier `toString`, but it's the _same_ item.

You get that second case if the `toString` value is e.g. `getSystemName()+" ("+getUserName()+")"` or similar where the `getSystemName()` at the front provides enough for a sort.  (There's still an issue over parsable, which includes that sub-case of sortable, system names, but we're making progress consistent on that)

Type II has some additional issues in our _existing_ code, where a lot of stuff is still passing around a `String` name instead of a bean references or a `NamedBeanReference`. But in the long term we need to make progress on that in any case, so I don't view that as an issue that would prevent a type-II `toString`.

Bottom line: I think we could standardize on permitting user name info in the NamedBean implementation of `toString()`.  It's not formally correct, but maybe the case can be made that it's OK if proceeded by the system name.  I think that additional state information could also be OK (again preceded by system name), but that's pretty far from the common definition, and I'd really like for somebody to make an argument as to why that's a _necessary_ thing to include.

pabender commented:
@bobjacobsen I am not arguing that the current implementation of toString in IDTag and RailCom objects is correct.

What I am arguing is that the current implementation of reporters has forced the current behavior on us to get the result we want.

That particular result is that we want the user ID of the tag ( which typically is a reporting mark for a piece of equipment ) to appear in the report.  The IDTag implementation of toString does this by providing the userid or the tag ID.  This is all ok in the sense that unless there is an error, you don’t tend to change the reporting marks.  State information is never returned by the toString method.

The RailCom implementation of toString incorrectly provides state information from the report followed by the DCC address from the report.  The address itself is OK, but it really should at least follow the same pattern as the IDTags.  ( I did not write either of these implementations and have just lived with it to now ).

I am going to take on some work to fix a few issues in the Reporter mechanisms.  Here are the issues I want to address now:

First, we have Operations setup so that when an IDTag associated with a piece of rolling stock appears at a reporter, and the reporter is associated with a location, the location of that piece of rolling stock is updated to to match.  This works with both IDTags and RailCom objects ( which are derived from IDTags ).  It doesn’t work when the reporter reports a string object ( which is the case for LocoNet, RPS, and JMRIClient reporters ) so we need to look into fixing that when possible by changing the report types of the reporter to be something derived from IDTag.

Second, at least when the report is of a type we can control, the report text shown in a reporter on a panel or in the reporter table should not be the toString of the object, but a designated report string.  My thought is perhaps we should add a new reportable interface that defines a toReportString method.  We can then rewrite the reporter table code and the on panel reporter icons to use this method if available but use toString otherwise ( maintaining backward compatibility).

I will work on the second issue first.  Once this code is in place, IDTag and RailCom objects should be able to use a common toString.

bobjacobsen commented:
@pabender  I don't understand all the ramifications, but what you describe sounds right to me. Thank you.

Returning to the question of this PR, perhaps there's another approach (haven't tried this, might be subtle compiler issues).  

- We could provide a `default` implementation (which can't be `final`, but doesn't need to be) that's `getSystemName()+" ("+getUserName()+")"` (just `getSystemName()` would be fine too) that's annotated with `@OverridingMethodsMustInvokeSuper` and is documented to say that the `super.toString()` content has to be the _first_ part of anything that a sub-class returns.

- We don't have any static checks that enforce `@OverridingMethodsMustInvokeSuper`(yet), but we could manually go through and make it true.

- Then, if we get artifacts because e.g. something is storing `toString()` instead of a `NamedBeanreference` (i.e. there's a problem when a user name is changed), we just fix those in the appropriate way.

I think there is some value in at least getting to a clearly-documented understanding that `toString()` has to start with system name.

danielb987 commented:
See PR #6464 for a possible solution on the LocoNet problem.

danielb987 commented on java/src/jmri/jmrix/loconet/TranspondingTag.java:
```diff
@@ -32,7 +32,7 @@ public TranspondingTag(String systemName, String userName) {
 
     @Override
-    public String toString(){
+    public String toReportString(){
```
If I revert this change, JMRI will not compile since this PR makes AbstractNamedBean.toString() final and therefore TranspondingTag cannot override toString().

The purpose of this PR is to show the consequences of making `AbstractNamedBean.toString()` final.

I can revert the change if the purpose of doing that is to get the compile error.

pabender replied:
There are two problems here:

1. This isn’t the right change.  The intention is that the toString in TranspondingTag eventually will just be removed.
2. We need to give users time to adjust ( scripts in particular).  

In another PR, I added the proper deprecation warnings to the toString method in TranspondingTag.  Our normal cycle for removing those is at least 2 years.

danielb987 replied:
The change is now reverted.

danielb987 commented on java/test/jmri/jmrix/loconet/TranspondingTagTest.java:
```diff
@@ -40,13 +40,15 @@ public void testTranspondingTagToString() {
         // set the entryexit property
         r.setProperty("entryexit","exits");
-        Assert.assertEquals("TranspondingTag toString ", "1234 exits", r.toString());
+        Assert.assertEquals("TranspondingTag toString ", "ID1234", r.toString());
```
The change is now reverted.

danielb987 commented:
Due to `jmri.jmrix.loconet.TranspondingTag.toString()` is marked depricated but will be kept for maybe two years, this PR will not be able to be completed until then.

pabender commented:
@danielb987 a minor correction to my earlier statement, our documentation says we keep deprecated things used in scripts around for at least two production releases:

http://jmri.org/help/en/html/doc/Technical/RP.shtml

So a year from now we can remove that.

(and yes, I still need to update the sample scripts that use this)

rhwood commented:
Making `AbstractNamedBean.toString()` final removes apparently valuable effectively constant data from the toString() descriptor of overriding methods (especially in that some Audio-specific NamedBeans's toString() method indicates ability to be used for purpose in logs), so if you insist on making this method final, please ensure that all logging statements do not loose information as a result.

danielb987 commented:
This PR needs to wait until next major release due to removal of depricated methods.

Also the [comment](https://github.com/JMRI/JMRI/pull/6463#issuecomment-507604333) by @rhwood needs to be resolved.

bobjacobsen commented:
A year has gone by, so it seems time to revive this for 4.19.1.  What else needs to be done?

danielb987 commented:
I have added the method `AudioSource.getDebugString()` which can be used to get more debug information. This method is overridden by `AbstractAudioSource`.

I have updated the `log` statements in `JavaSoundAudioSource` to use `AudioSource.getDebugString()` and I don't think there is any more log statements in the `jmri.jmrit.audio` package that should use it. But there may be cases outside this package and I'm not sure how to find those.

danielb987 commented:
> @danielb987 a minor correction to my earlier statement, our documentation says we keep deprecated things used in scripts around for at least two production releases:
> 
> http://jmri.org/help/en/html/doc/Technical/RP.shtml
> 
> So a year from now we can remove that.
> 
> (and yes, I still need to update the sample scripts that use this)

@pabender have you updated the sample scripts? Is this ready for you?

danielb987 commented:
> A year has gone by, so it seems time to revive this for 4.19.1. What else needs to be done?

I think it's ready to be merged, but it would be good if @pabender and @rhwood can take a look at it first to check if it's ready for them.

danielb987 commented:
I get these errors on Travis

danielb987 commented:
I get these errors on Travis HEADLESS=false SKIPINTERMITTENT=true STATIC=false but they are unrelated to this PR:

```
[ERROR] Failures: 
[ERROR] jmri.configurexml.LoadAndStoreTest.loadLoadStoreFileCheck[java/test/jmri/configurexml/load/LeadingTrailingSpaceTest.xml (pass=true)](jmri.configurexml.LoadAndStoreTest)
[ERROR]   Run 1: LoadAndStoreTest>LoadAndStoreTestBase.loadLoadStoreFileCheck:295->LoadAndStoreTestBase.checkFile:213 expected:<  <[signalmastlogics class="jmri.managers.configurexml.DefaultSignalMastLogicManagerXml"]>> but was:<  <[filehistory]>>
[ERROR]   Run 2: LoadAndStoreTest>LoadAndStoreTestBase.tearDown:314 Unexpected ERROR or higher messages emitted:          to file2:"/home/travis/build/JMRI/JMRI/temp/temp/LeadingTrailingSpaceTest.xml"
```

rhwood commented:
`Object.toString()` is useful for debugging (and commonly used only for that purpose) to the extent that all modern logging frameworks are designed to use `toString()` to make (not) logging inexpensive by not having to invoke a method against the object being logged until after the decision to log has been made.

A quick review of the current persistence code suggests that `toString()` is not used for NamedBeans, and a similar review suggests that it `toString()` is not used to identify a NamedBean in the tables. I think most GUI code that sorts NamedBeans now does so correctly without resorting to using `toString()`, so the original intent of this PR is lost.

Basically what I am getting at is that maybe (especially since `toString()` cannot be made final in `jmri.NamedBean`), maybe (although I generally *like* the changes in here), the requirement that it be final in `jmri.implementation.AbstractNamedBean` be dropped, and the documentation comment for `toString()` should be changed to reflect the ongoing reality that it is not always going to be just the same result as `getSystemName()`.

So I’d like to see two things:
- [ ] remove final modifier from `AbstractNamedBean.toString()`
- [ ] revert changes to AbstractAudio to allow `toString()` to provide more information

Otherwise this generally a good code cleanup.

danielb987 commented:
@bobjacobsen @pabender 
What is your thoughts about the comment of @rhwood above? What is the best way forward?
There is no need to address this before the holidays. I will wait for your comments before taking any action on this.

danielb987 commented:
@bobjacobsen 
Please be careful when merge this branch into master or development. See for example `help/en/JavaHelpSearch/SCHEMA` which wants to change `id1=17175` to `id1=17172`. This may be due to that this branch wants to merge into JMRI:development.

bobjacobsen commented:
I'm switching this back to `master` as part of returning to our previous process.

danielb987 commented:
> I'm switching this back to `master` as part of returning to our previous process.

Thanks.

danielb987 commented:
> `Object.toString()` is useful for debugging (and commonly used only for that purpose) to the extent that all modern logging frameworks are designed to use `toString()` to make (not) logging inexpensive by not having to invoke a method against the object being logged until after the decision to log has been made.
> 
> A quick review of the current persistence code suggests that `toString()` is not used for NamedBeans, and a similar review suggests that it `toString()` is not used to identify a NamedBean in the tables. I think most GUI code that sorts NamedBeans now does so correctly without resorting to using `toString()`, so the original intent of this PR is lost.
> 
> Basically what I am getting at is that maybe (especially since `toString()` cannot be made final in `jmri.NamedBean`), maybe (although I generally _like_ the changes in here), the requirement that it be final in `jmri.implementation.AbstractNamedBean` be dropped, and the documentation comment for `toString()` should be changed to reflect the ongoing reality that it is not always going to be just the same result as `getSystemName()`.
> 
> So I’d like to see two things:
> 
> * remove final modifier from AbstractNamedBean.toString()
> * revert changes to AbstractAudio to allow toString() to provide more information
>
> Otherwise this generally a good code cleanup.

@bobjacobsen @pabender
What is your thoughts about the comment of @rhwood above? What is the best way forward?

bobjacobsen commented:
To me, this PR seems to be more about "have a contract on the content of NamedBean#toString()" rather than about it being final.  

The general Object#toString() contract doesn't really help here:
> Returns a string representation of the object.

The literature shows that there are two basic opinions for what toString should provide:  The identity of the object, or the content.  

 - The default toString in the Object class is clearly providing the identity: It shows the address of the object.
 - Many complex Java-provided objects, particularly in Swing, provide the content: They show a long list of values of internal variables

The middle ground is populated a little:  Some Swing components expect that toString will provide the string that should be shown in the GUI, for example, which probable is a representation of just some of the content in a nicely-formatted form.

1. I think the right thing for toString() of a NamedBean to return is it's SystemName. If somebody were to give a good reason for some subclass to do more than that, that's fine, I just can't think of one now.  But I think there should be a strong convention that the SystemName is always there.

2. Randall has a good point that Java itself doesn't provide a way to insist that an implementation in NamedBean itself is final (default methods can't be final).  If that's an important thing to accomplish, we could move the common toString implementation up to NamedBean, add a strong comment, and [add an architecture check](https://www.archunit.org/userguide/html/000_Index.html#_composing_member_rules) that ensures that nothing in our code overrides it.  Alternately, we can leave it in AbstractNamedBean as final, and then add a check that insists that any implementation of NamedBean goes through that. Either way means anybody else is a bit on their own, but that's always true.

danielb987 commented:
Following @bobjacobsen comments:
This PR now does (1) by making AbstractNamedBean.toString() final and return system name for any NamedBean that extend AbstractNamedBean.

If someone instead wants (2), it would be easy to move toString() to NamedBean and add the architecture check after this PR is merged.

So this PR is ready to be merged.

bobjacobsen commented:
@danielb987 Thank you for your work on this!  

I'm going to close/reopen to get a clean set of CI runs. Once that runs clean, we can merge.

geowar1 commented:
If .toString can only return the system name then why have two API's? If I want the system name I'll ask for it. The .toString method should NEVER be used for GUI (explicit methods /formatting, etc. should always be used instead). So that pretty much leaves .toString for debugging use. When I mouse over a variable I expect to see the .toString value (including any variable state that the .toString developer for that class felt compelled to include). There is no contract (implicit or otherwise) that says .toString can't include variable state and a lot of implementations that do exactly that. IMHO AbstractNamedBean should include the system name AND the .toString of the object that it's the named bean of; and that about it.

danielb987 commented:
There are 12 classes there the method toString() is removed by this PR. 8 of these belongs to Audio. Two other is DefaultSignalSystem and TranspondingTag.

The last two is AbstractAudio and AbstractStringIO which only returns the name of the class and the system name, and therefore doesn't give any more value than the current implementation.

So there is very little use of toString() as a debug tool inside JMRI. For AudioSource, I have replaced the method `toString()` with the method `getDebugString()`.

rhwood commented:
> So there is very little use of toString() as a debug tool inside JMRI. For AudioSource, I have replaced the method toString() with the method getDebugString().

`toString()` is used in *every* `log.debug(...)` call in JMRI that includes the String representation of objects--its used extensively for debugging within JMRI.

danielb987 commented:
What I meant was that there is only 11 classes that provides their own implementation. And two of them doesn't do any useful, they only return name of class + system name.

Besides Audio, the only two other classes that provides any valuable additional information is DefaultSignalSystem and TranspondingTag.

pabender commented:
> On Jan 19, 2020, at 9:37 PM, Daniel <notifications@github.com> wrote:
> Besides Audio, the only two other classes that provides any valuable additional information is DefaultSignalSystem and TranspondingTag.
> 

The TranspondingTag toString method was implemented the way it was for compatibility with scripts written with the prior Transponding implementation, which returned a string.  It was deprecated upon addition to give users time to transition to using the toReportString method of the reportable interface instead.

Ultimately, I do want the toString for all the IdTag objects to use the same implementation, so removing the special case from TranspondingTag at this point is fine.

geowar1 commented:
>>What I meant was that there is only 11 classes that provides their own implementation.

cd java/src
find . -type f -name '*.java' -exec grep -li 'String toString' {} \;|wc -l
189

>> getDebugString

Will that be used instead of toString when mousing over variables in NetBeans?

rhwood commented:
No, NetBeans cannot use `getDebugString()` when mousing over variables because that method is not part of the Java API.

danielb987 commented:
CI has now completed successfully.

danielb987 commented:
This PR is ready to be merged. AppVeyor fails, but that is probably not related to this PR.

bobjacobsen commented:
Reran tests: Travis clear, AppVeyor two different well-known intermittent failures
